### ISSUE TITLE
Does explainer.shap_values preserve the original order of the dataset?

### ISSUE THREAD

**Issue by palpen** [shap/shap#1250]
I'm using the SHAP library to explain predictions of a credit risk model. 
I need to concatenate the customer IDs back to the SHAP values generated by the explainer.shap_values method (the customer ID was never used to train the model). 

If I were to convert the SHAP value numpy array to a dataframe with 

    shap_df = pd.DataFrame(shap_values[1])
    shap_df.columns = X.drop('customer_id', 1).columns

is it safe to do

    shap_df['customer_id'] = X['customer_id']

?

In other words, does the explainer.shap_values method preserve the original order of the dataset? If not, what is the recommended approach to concatenate back the customer IDs?

Thank you

**Comment by palpen** [comment]
I just did a quick experiment creating a pseudo-customer ID using the index of the adult income data from the documentation. It doesn't look like order is preserved:

```
import lightgbm as lgb
import shap
import pandas as pd

X_adult, y_adult = shap.datasets.adult(display=True)
X_adult = X_adult.reset_index()  # pseudo-Customer IDs
X_adult = X_adult.rename({'index':'cust_id'})
cat_cols = ['Workclass', 'Marital Status', 'Occupation', 'Relationship', 'Race', 'Sex', 'Country']
d_train = lgb.Dataset(X_adult.drop(columns=['cust_id']), 
                                     label=y_adult,
                                     categorical_feature=cat_cols)

params = {'boosting_type': 'gbdt',
          'max_depth' : -1,
          'objective': 'binary',
          'nthread': 3, # Updated from nthread
          'num_leaves': 64,
          'learning_rate': 0.05,
          'max_bin': 512,
          'subsample_for_bin': 200,
          'subsample': 1,
          'subsample_freq': 1,
          'colsample_bytree': 0.8,
          'reg_alpha': 5,
          'reg_lambda': 10,
          'min_split_gain': 0.5,
          'min_child_weight': 1,
          'min_child_samples': 5,
          'scale_pos_weight': 1,
          'num_class' : 1,
          'metric' : 'binary_error'}
model = lgb.train(train_set=d_train, params=params)

explainer = shap.TreeExplainer(model)
shap_values = explainer.shap_values(X_adult.drop(columns=['cust_id']))

shap_df = pd.DataFrame(shap_values[1])
shap_df.columns = X_adult.drop('cust_id', 1).columns
shap_df['cust_id'] = X_adult['cust_id']

cust_id = 2
display(X_adult[X_adult.cust_id == cust_id].T)
shap.decision_plot(explainer.expected_value[1], 
                   shap_values[1][cust_id], 
                   X_adult.drop('cust_id', 1), 
                   feature_order="None")
```

Looking at customer ID ('cust_id') number 2, the record in the feature dataset is a white male age 38 with 0 capital gains. In the decision plot, however, we have a white male age 39 with 2,174 capital gains.

<img width="530" alt="Screen Shot 2020-06-03 at 8 42 47 PM" src="https://user-images.githubusercontent.com/9051141/83702480-d26c7c00-a5da-11ea-80b5-3af1232b12ad.png">

**Comment by palpen** [comment]
Nevermind. Turns out I'm not selecting the correct record in

```
cust_id = 2
display(X_adult[X_adult.cust_id == cust_id].T)
shap.decision_plot(explainer.expected_value[1], 
                   shap_values[1][cust_id], 
                   X_adult.drop('cust_id', 1), 
                   feature_order="None")
```

I should also select the same cust_id in X_adult

```
cust_id = 2
display(X_adult[X_adult.cust_id == cust_id].T)
shap.decision_plot(explainer.expected_value[1], 
                   shap_values[1][cust_id], 
                   X_adult.drop('cust_id', 1).iloc[cust_id,:], 
                   feature_order="None")
```
